### PR TITLE
Script to run all plugins tests and try an install, rooting in local Logstash

### LINKED ISSUE #15160: Create a script to unit test all supported plugins

### PR DESCRIPTION
**Pull request by andsel**
## Release notes

[rn:skip]

## What does this PR do?

Implements a script to test all supported plugins against Logstash/JRuby, in particular it uses the JRuby(and JDK) bundled with Logstash to execute the unit tests, create the gem and install the gem on local Logstash.

This doesn't start a real pipeline run into the local Logstash, but should be enough to reveal macroscopic Ruby syntax compatibility errors.

## Why is it important/What is the impact to the user?

The introduced script permit to test supported plugins in the context of Logstash. 
It provides a couple of command line options to select the plugins to test.
Using `-p` followed by a plugin name verify only that plugin. To test `logstash-integration-jdbc` it can be specified as:
```sh
ci/test_supported_plugins.sh -p logstash-integration-jdbc
```

Using  `-t` (tier) and `-k` (kind) a set of plugins can be selected, for example: all input and output plugins of tier1 and tier2 could expressed as
```sh
ci/test_supported_plugins.sh -k input,output -t tier1,tier2
```

`-h` (halt) option is a boolean switch to stop on the first error instead of continue with all the plugins.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files (and/or docker env variables)~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~

## Author's Checklist

- [ ]

## Related issues

- Relates #14996
- Fixes #15160

### CODE REVIEW

**Review by yaauie — COMMENTED** [review]
I think this will be super useful, and spent a bit of time bike-shedding where the list of supported plugins should be (I don't believe the current nested structured embedding is the "right" place).

I ran into a bit of a problem though because this script pollutes the `Gemfile` and `Gemfile.lock` files that are used by Logstash's plugin manager, and the `ci/test_supported_plugins.sh` entry-point invokes `./gradlew installDefaultGems` which ensures that most (all?) of these gems are already previously installed from their relevant rubygems artifact.

Questions to that end:

1. do we expect this to run _on top of_ the default gem installation, and if so, what are we doing to validate that the local-git-checkout of the plugin is what is effectively installed?
2. are we trying to test installing each in isolation (that is, snapshot the `Gemfile` and `Gemfile.lock` before each plugin we are validating and restore the snapshot before proceeding to the next), or are we trying to test installing them together (if so, we should likely randomize the order in which we do so)

File: ci/test_supported_plugins.rb [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    puts "gem_file generated: #{gem_file}"
+    gem_file
+  end
+
+  def install_gem(gem_file)
+    logstash_plugin_cli = ENV['LOGSTASH_PATH'] + "/bin/logstash-plugin"
+    stdout, stderr, status = Open3.capture3("#{logstash_plugin_cli} install #{gem_file}")
+    reg_exp = /Installing .*\nInstallation successful$/
+    if status != 0 && !reg_exp.match(stdout)
+      puts "Failed to install plugins:\n #{stdout}"
```
[the remark this answers is not in the document]
Comment (andsel) [reply ci/test_supported_plugins.rb]: Fixed with: 5942f577c981ffaad55383214123381b1078d5be

**Comment by andsel** [comment]
Thanks @yaauie  for reviewing it!

> I ran into a bit of a problem though because this script pollutes the Gemfile and Gemfile.lock files that are used by Logstash's plugin manager, and the ci/test_supported_plugins.sh entry-point invokes ./gradlew installDefaultGems which ensures that most (all?) of these gems are already previously installed from their relevant rubygems artifact.

Ahh right, but the `bundle install` executed during the testing phase shouldn't overwrite with the latest dependencies required by the plugin itself? https://github.com/elastic/logstash/pull/15018/files#diff-a23b8397e5c96e0774b4b531cb29933248fb8278670d575b0acd4a5575985770R59 in that case it should test the plugin with the right dependencies, I think.

> 1. do we expect this to run on top of the default gem installation, and if so, what are we doing to validate that the local-git-checkout of the plugin is what is effectively installed?

You are right, maybe before installing the locally created gem the script should remove the plugin from Logstash with a `bin/logstash-plugin uninstall <plugin pre-installed>`, WDYT, coule be enough?

> 2. are we trying to test installing each in isolation..., or are we trying to test installing them together
Actually the script install one after the other without isolation of installs a "copy & restore" of `Gemfile` and `Gemfile.lock` would help a lot into this.

File: ci/test_supported_plugins.rb [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+  def initialize(plugins_folder, plugin_name)
+    @plugin_name = plugin_name
+    @plugins_folder = plugins_folder
+    @plugin_base_folder = "#{plugins_folder}/#{plugin_name}"
+  end
+
+  def git_clone
+    if File.directory?(plugin_name)
+      puts "#{plugin_name} already cloned locally"
+      return
```
Comment (yaauie) [inline ci/test_supported_plugins.rb:41]: should we `git pull` in this case? do we need to worry about "dirty" repositories with untracked or modified files?

File: ci/test_supported_plugins.rb [hunk 3]
```diff
@@ -0,0 +1,253 @@
+# encoding: utf-8
+
+# Test the 
+# - build (rspec), 
+# - packaging (gem build)
+# - and deploy (bin/logstash-plugin install)
+# of a plugins inside the current Logstash, using its JRuby
+# Usage example:
+# bin/ruby ci/test_supported_plugins.rb -p logstash-integration-jdbc
+# bin/ruby ci/test_supported_plugins.rb -t tier1 -k input,codec,integration
```
Comment (yaauie) [inline ci/test_supported_plugins.rb:10]: 🤔 the usage is missing a note about needing a clean bootstrap + `installDevelopmentGems`.

**Review by yaauie — COMMENTED** [review]
Excellent.

It would be good to have a singular canonical reference of plugins by support tier that can be referenced by this script instead of hard-coding it, but that can be deferred to a separate effort.

File: ci/test_supported_plugins.rb (hunk 2 above)
Comment (andsel) [reply ci/test_supported_plugins.rb:53]: Yes, great idea thanks. Implemented with commit e23677bbbaf2baa2c7e6372f3985bdaa44c38854

File: ci/test_supported_plugins.rb [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+require 'optparse'
+
+ENV['LOGSTASH_PATH'] = Dir.pwd
+ENV['LOGSTASH_SOURCE'] = '1'
+
+logstash_plugin_cli = ENV['LOGSTASH_PATH'] + "/bin/logstash-plugin"
+
+# it has to be out of logstash local close else plugins' Gradle script
+# would interfere with Logstash's one
+base_folder = "/tmp"
```
[the remark this answers is not in the document]
Comment (andsel) [reply ci/test_supported_plugins.rb]: Good point, fixed with commit d3df637c5472408d4386b1dd4391b11229d20095

File: ci/test_supported_plugins.rb (hunk 3 above)
Comment (andsel) [reply ci/test_supported_plugins.rb:10]: we need the Logstash JRuby to be present, so at least here we should execute `./gradlew assemble`, to setup the environment to be able to run the Ruby script.

We can split this in 2:
- in the shell script invoke the minimal Gradle task that setup the environment to be able to run the Ruby script
- in the Ruby script executes all the steps that are fundamental to have Logstash in right posture to execute the tests; which means executing the Gradle's `installDevelopmentGems` tasks

Comment (andsel) [reply ci/test_supported_plugins.rb:10]: handled with b30ad380adb9cf70a039c2752fadb44625a0865b

**Comment by andsel** [comment]
Hi @yaauie 
> It would be good to have a singular canonical reference of plugins by support tier that can be referenced by this script instead of hard-coding it, but that can be deferred to a separate effort.

I agree with you. My idea would be to move those tiers and plugins inventory into a separate YAML file, maybe in the root of the Logstash project.

A sample could be:
```yaml
tier1:
  inputs:
    - logstash-input-azure_event_hubs
    - logstash-input-beats
  codecs:
   - logstash-codec-generator
  filters:
    - ...
  outputs:
   - ...
  integrations:
   - logstash_integration_kafka

tier2: ...    
```

I think that would be better addressed in a follow up PR. WDYT?

**Comment by andsel** [comment]
Eventually we could use this: https://github.com/elastic/logstash/blob/main/rakelib/plugins-metadata.json as plugin repository. Every plugins that has `default-plugins` to true is bundled in Logstash default distribution, so it's supported plugin.

**Review by yaauie — APPROVED** [review]
Implementation LGTM.

Let's merge as-is file an issue about the shared source-of-truth for plugins and their tiers so we don't lose track of it.

**Comment by andsel** [comment]
Created the follow up issue #15190 to track the request to use a common plugins inventory.

**Pull request merged**